src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs

```diff
@@ -1885,139 +2021,186 @@ private static BigInteger Subtract(uint[]? leftBits, int leftSign, uint[]? right
-                uint xu = (i < x.Length) ? x[i] : xExtend;
-                uint yu = (i < y.Length) ? y[i] : yExtend;
+                uint xu = ((uint)i < (uint)x.Length) ? x[i] : xExtend;
+                uint yu = ((uint)i < (uint)y.Length) ? y[i] : yExtend;
                 z[i] = xu ^ yu;
             }
 
-            return new BigInteger(z);
+            if (leftBufferFromPool != null)
+                ArrayPool<uint>.Shared.Return(leftBufferFromPool);
+
+            if (rightBufferFromPool != null)
+                ArrayPool<uint>.Shared.Return(rightBufferFromPool);
+
+            var result = new BigInteger(z);
+
+            if (resultBufferFromPool != null)
+                ArrayPool<uint>.Shared.Return(resultBufferFromPool);
+
+            return result;
         }
 
         public static BigInteger operator <<(BigInteger value, int shift)
         {
-            if (shift == 0)
-                return value;
+            if (shift == 0) return value;
+            else if (shift == int.MinValue) return ((value >> int.MaxValue) >> 1);
+            else if (shift < 0) return value >> -shift;
 
-            if (shift == int.MinValue)
-                return (value >> int.MaxValue) >> 1;
+            int digitShift = shift / kcbitUint;
+            int smallShift = shift - (digitShift * kcbitUint);
 
-            if (shift < 0)
-                return value >> -shift;
+            uint[]? xdFromPool = null;
+            int xl = value._bits?.Length ?? 1;
+            Span<uint> xd = (xl <= BigIntegerCalculator.StackAllocThreshold ?
+                            stackalloc uint[BigIntegerCalculator.StackAllocThreshold]
+                            : xdFromPool = ArrayPool<uint>.Shared.Rent(xl)).Slice(0, xl);
+            bool negx = value.GetPartsForBitManipulation(xd);
 
-            (int digitShift, int smallShift) = Math.DivRem(shift, kcbitUint);
+            int zl = xl + digitShift + 1;
+            uint[]? zdFromPool = null;
+            Span<uint> zd = ((uint)zl <= BigIntegerCalculator.StackAllocThreshold ?
+                            stackalloc uint[BigIntegerCalculator.StackAllocThreshold]
+                            : zdFromPool = ArrayPool<uint>.Shared.Rent(zl)).Slice(0, zl);
+            zd.Clear();
 
-            Span<uint> xd = stackalloc uint[1];
-            bool negx = GetPartsForBitManipulation(ref value, ref xd);
-
-            int zl = xd.Length + digitShift + 1;
-            uint[]? zdArray = null;
-            Span<uint> zd = stackalloc uint[0];
-            if (zl <= StackallocUInt32Limit)
-            {
-                zd = stackalloc uint[StackallocUInt32Limit].Slice(0, zl);
-                zd.Slice(0, digitShift).Clear();
-            }
-            else
-            {
-                zd = zdArray = new uint[zl];
-            }
-
-            uint carry = 0;
             if (smallShift == 0)
             {
-                for (int i = 0; i < xd.Length; i++)
+                for (int i = 0; i < xl; i++)
```

Comment:
Prefer `xd.Length` over `xl` here. More readable and JIT recognizes this pattern for bounds check elision. (Same comment line 2159 below.)

Reply:
Fixed and reviewed across entire code base.